otel-extension/src/main/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ConcurrentMetricRegistryPoller.scala

```diff
@@ -0,0 +1,60 @@
+package io.scalac.mesmer.otelextension.instrumentations.zio
+
+import java.util.Timer
+import java.util.TimerTask
+
+import zio.metrics.MetricKey
+import zio.metrics.MetricKeyType
+
+import scala.collection.mutable
+import scala.concurrent.duration._
+import scala.util.control.NonFatal
+
+import io.scalac.mesmer.otelextension.instrumentations.zio.ZIOMetrics._
+
+class ConcurrentMetricRegistryPoller {
+
+  private val timer = new Timer()
+
+  private val instruments = mutable.HashMap.empty[MetricKey[MetricKeyType], AutoCloseable]
```

Comment:
Looking at the VirtualFIeld docs now: 

```
/**
 * Represents a "virtual" field of type {@code F} that is added to type {@code T} in the runtime.
 *
 * <p>A virtual field has similar semantics to a weak-keys strong-values map: the value will be
 * garbage collected when their owner instance is collected. It is discouraged to use a virtual
 * field for keeping values that might reference their key, as it may cause memory leaks.
 *
 * @param <T> The type that will contain the new virtual field.
 * @param <F> The field type that'll be added to {@code T}.
 */
 ```
 
 It makes me wonder if we can use it for the instruments and have the resources cleaned by the VF itself (when MetricKey is collected, the instrument is as well?)

Reply:
There is an `Autocloseable` that also needs to be closed, so garbage collection is not enough - we also need to execute a callback. 

I guess we could also fully tie this functionality to `MetricKey` lifecycle - setup instruments in their constructor, tear down on garbage collection or something like that. This would remove the need for polling. Though I do not think the polling is overly expensive, since it just processes an internal map. But maybe such approach would be somewhat cleaner?

Reply:
> Autocloseable 

Oh right.

Let's stick with the current solution and then iterate on it if we find it annoying/wrong. For now it's more than enough